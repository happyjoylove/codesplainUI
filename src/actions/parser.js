export const LOAD_PARSER = 'LOAD_PARSER';
export const ADD_ERROR = 'ADD_ERROR';
export const CLEAR_ERRORS = 'CLEAR_ERRORS';

export const loadParser = language => ({
  type: LOAD_PARSER,
  meta: {
    WebWorker: true,
  },
  payload: language,
});

export const addError = error => ({
  type: ADD_ERROR,
  payload: error,
});

export const clearErrors = () => ({
  type: CLEAR_ERRORS,
});
