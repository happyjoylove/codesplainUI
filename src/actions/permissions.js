export const SET_PERMISSIONS = 'SET_PERMISSIONS';
export const SET_AUTHOR = 'SET_AUTHOR';

export const setPermissions = permissions => ({
  type: SET_PERMISSIONS,
  payload: permissions,
});

export const setAuthor = author => ({
  type: SET_AUTHOR,
  payload: author,
});
