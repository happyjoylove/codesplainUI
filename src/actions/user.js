import axios from 'axios';
import mapValues from 'lodash/mapValues';

import { makeSaveEndpointUrl, sanitizeSnippetList } from '../util/requests';

const API_URL = process.env.REACT_APP_API_URL;

// Util func to check for 'NoSuchKey' responses from S3
const noSuchKey = data => (
  typeof data === 'string' && data.includes('NoSuchKey')
);

export const ADD_ORG = 'ADD_ORG';
export const ADD_ORGANIZATIONS = 'ADD_ORGANIZATIONS';
export const CLEAR_USER_CREDENTIALS = 'CLEAR_USER_CREDENTIALS';
export const RESTORE_USER_CREDENTIALS = 'RESTORE_USER_CREDENTIALS';
export const SAVE_ACCESS_TOKEN = 'SAVE_ACCESS_TOKEN';
export const SAVE_USERNAME = 'SAVE_USERNAME';
export const SET_AVATAR_URL = 'SET_AVATAR_URL';
export const SET_SNIPPET_LISTS = 'SET_SNIPPET_LISTS';
export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const SWITCH_ORG = 'SWITCH_ORG';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';

export const setAvatarUrl = url => ({
  type: SET_AVATAR_URL,
  payload: url,
});

export const addOrg = org => ({
  type: ADD_ORG,
  payload: org,
});

export const addOrganizations = organizations => ({
  type: ADD_ORGANIZATIONS,
  payload: organizations,
});

export const clearUserCredentials = () => ({
  type: CLEAR_USER_CREDENTIALS,
});

export const saveAccessToken = accessToken => ({
  type: SAVE_ACCESS_TOKEN,
  payload: accessToken,
});

export const saveUsername = username => ({
  type: SAVE_USERNAME,
  payload: username,
});

export const setSnippetLists = snippetLists => ({
  type: SET_SNIPPET_LISTS,
  payload: snippetLists,
});

export const setUserSnippets = snippetMeta => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const switchOrg = org => ({
  type: SWITCH_ORG,
  payload: org,
});

export const updateUserSnippets = () => (dispatch, getState) => {
  const {
    user: {
      token,
      username,
    },
  } = getState();
  // Fetch the user's snippet meta data and save it
  const headers = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios.get(makeSaveEndpointUrl(username), { headers })
    .then(({ data }) => {
      // Jump to catch block if the user has no index.json file:
      if (noSuchKey(data)) {
        throw new Error(`index.json does not exist for ${username}!`);
      }
      dispatch(setUserSnippets(data));
    });
};

export const fetchSnippetLists = () => (dispatch, getState) => {
  const {
    token,
    orgs,
  } = getState().user;
  const headers = {
    Authorization: `token ${token}`,
  };
  const url = `${API_URL}/users?users=${orgs.join(',')}`;
  // Remove the username from the org list
  return axios({
    method: 'GET',
    url,
    headers,
  })
    .then(({ data: snippetLists }) => {
      const sanitizedSnippetLists = mapValues(snippetLists, sanitizeSnippetList);
      dispatch(setSnippetLists(sanitizedSnippetLists));
    });
};

export const fetchAccessToken = authCode => () => {
  const reqUrl = `${API_URL}/auth`;
  const reqBody = { code: authCode };
  return axios({
    method: 'POST',
    url: reqUrl,
    data: reqBody,
  });
};

export const fetchUserInfo = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user',
    headers: reqHeaders,
  });
};

export const fetchUserOrgs = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user/orgs',
    headers: reqHeaders,
  });
};
