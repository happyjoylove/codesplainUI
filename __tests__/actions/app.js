import moxios from 'moxios';

import * as actions from '../../src/actions/app';
import generateMockStore from '../../src/testUtils/mockStore';

describe('Actions: App', () => {
  it('creates correct SET_SNIPPET_KEY action object', () => {
    const key = 'szechuan sauce';
    expect(actions.setSnippetKey(key)).toMatchSnapshot();
  });
  it('creates correct SET_SNIPPET_CONTENTS action object', () => {
    const snippet = 'Show me what you got';
    expect(actions.setSnippetContents(snippet)).toMatchSnapshot();
  });
  it('creates correct SET_RULE_FILTERS action object', () => {
    const filters = {
      for_stmt: {
        prettyTokenName: 'For Loops',
        count: 1,
        selected: false,
      },
      atom: {
        prettyTokenName: 'Atoms',
        count: 21,
        selected: true,
      },
    };
    expect(actions.setRuleFilters(filters)).toMatchSnapshot();
  });
  it('creates correct RESET_RULE_FILTERS action object', () => {
    expect(actions.resetFilters()).toMatchSnapshot();
  });
  it('creates correct SELECT_ALL_FILTERS action object', () => {
    expect(actions.selectAllFilters()).toMatchSnapshot();
  });
  it('creates correct RESET_STATE action object', () => {
    expect(actions.resetState()).toMatchSnapshot();
  });
  it('creates correct RESET_STATE action object', () => {
    const AST = {
      type: 'file_input',
      begin: 0,
      end: 1,
      children: [
        'NEWLINE',
        'INDENT',
        'NEWLINE',
        'DEDENT',
      ],
    };
    expect(actions.setAST(AST)).toMatchSnapshot();
  });
  it('creates correct TOGGLE_EDITING_STATE action object', () => {
    expect(actions.toggleEditState()).toMatchSnapshot();
  });
  it('creates correct SAVE_ANNOTATION action object', () => {
    const annotationData = {
      annotation: 'You pass butter.',
      lineNumber: 1,
      lineText: 'What is my purpose?',
    };
    expect(actions.saveAnnotation(annotationData)).toMatchSnapshot();
  });
  it('creates correct RESTORE_STATE action object', () => {
    const savedState = {
      snippet: '\t',
      snippetTitle: '',
      annotations: {},
      AST: {
        type: 'file_input',
        begin: 0,
        end: 1,
        children: [
          'NEWLINE',
          'INDENT',
          'NEWLINE',
          'DEDENT',
        ],
      },
      filters: {},
      readOnly: false,
    };
    expect(actions.restoreState(savedState)).toMatchSnapshot();
  });
  it('creates correct SET_SNIPPET_TITLE action object', () => {
    const snippetTitle = 'Get Schwifty';
    const expected = {
      type: actions.SET_SNIPPET_TITLE,
      payload: snippetTitle,
    };
    expect(actions.setSnippetTitle(snippetTitle)).toEqual(expected);
  });
  it('creates correct SET_SNIPPET_LANGUAGE action object', () => {
    const snippetLanguage = 'Gromflomite';
    expect(actions.setSnippetLanguage(snippetLanguage)).toMatchSnapshot();
  });
  it('creates correct PARSE_SNIPPET action object', () => {
    const snippet = 'Grasssss tastes bad';
    expect(actions.parseSnippet(snippet)).toMatchSnapshot();
  });
  it('creates correct SAVE_SUCCEEDED action object', () => {
    const expected = {
      type: actions.SAVE_SUCCEEDED,
    };
    expect(actions.saveSucceeded()).toEqual(expected);
  });
  describe('async actions', () => {
    beforeEach(() => {
      moxios.install();
    });
    afterEach(() => {
      moxios.uninstall();
    });
    describe('saveNew()', () => {
      it('dispatches correct actions if save succeeded', () => {
        const key = 'gazorpazorp';
        const org = 'rick-sanchez';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
            response: { key, status: '200' },
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '' },
        });
        return store.dispatch(actions.saveNew(org))
          .then((returnVal) => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
            expect(returnVal).toEqual(key);
          });
      });
      it('dispatches correct actions if save failed', () => {
        const org = 'rick-sanchez';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '' },
        });
        return store.dispatch(actions.saveNew(org))
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
    describe('deleteSnippet()', () => {
      it('dispatches the correct actions if the action succeeded', () => {
        const snippetKey = 'snippetKey';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
          });
        });
        const store = generateMockStore({
          user: {
            token: '',
            selectedOrg: '',
          },
        });
        return store.dispatch(actions.deleteSnippet(snippetKey))
          .then(() => {
            expect(store.getActions()).toMatchSnapshot();
          });
      });
      it('dispatches the correct actions if the action fails', () => {
        const snippetKey = 'snippetKey';
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const store = generateMockStore({
          user: {
            token: '',
            selectedOrg: '',
          },
        });
        return store.dispatch(actions.deleteSnippet(snippetKey))
          .catch(() => {
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
    describe('saveExisting()', () => {
      it('dispatches the correct actions if the action succeeded', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '', username: '' },
        });
        return store.dispatch(actions.saveExisting())
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });
      it('dispatches the correct actions if the action failed', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '', username: '' },
        });
        return store.dispatch(actions.saveExisting())
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
    describe('loadSnippet()', () => {
      it('dispatches the correct actions if the action succeeded', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
            response: JSON.stringify({
              snippet: 'snippet',
            }),
          });
        });
        const store = generateMockStore({
          app: {},
          user: { token: '' },
        });
        const snippetOwner = 'snippetOwner';
        const snippetKey = 'key';
        return store.dispatch(actions.loadSnippet(snippetOwner, snippetKey))
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
  });
});
