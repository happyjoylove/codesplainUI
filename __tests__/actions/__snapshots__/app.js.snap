exports[`Actions: App async actions deleteSnippet() dispatches the correct actions if the action fails 1`] = `
Array [
  Object {
    "payload": "Deleting...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Failed to delete snippet; please try again",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions deleteSnippet() dispatches the correct actions if the action succeeded 1`] = `
Array [
  Object {
    "payload": "Deleting...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Snippet Deleted!",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions loadSnippet() dispatches the correct actions if the action succeeded 1`] = `
Array [
  Object {
    "payload": Object {
      "AST": undefined,
      "annotations": Object {},
      "filters": Object {},
      "readOnly": false,
      "snippet": "snippet",
      "snippetLanguage": "python3",
      "snippetTitle": "title",
    },
    "type": "RESTORE_STATE",
  },
  Object {
    "payload": "key",
    "type": "SET_SNIPPET_KEY",
  },
  Object {
    "payload": "snippetOwner",
    "type": "SET_AUTHOR",
  },
]
`;

exports[`Actions: App async actions saveExisting() dispatches the correct actions if the action failed 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "payload": "Failed to save snippet; please try again",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions saveExisting() dispatches the correct actions if the action succeeded 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Codesplaination Saved!",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "SAVE_SUCCEEDED",
  },
]
`;

exports[`Actions: App async actions saveNew() dispatches correct actions if save failed 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "payload": "Failed to save snippet; please try again",
    "type": "ADD_NOTIFICATION",
  },
]
`;

exports[`Actions: App async actions saveNew() dispatches correct actions if save succeeded 1`] = `
Array [
  Object {
    "payload": "Saving...",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "CLOSE_NOTIFICATION",
  },
  Object {
    "payload": "Codesplaination Saved!",
    "type": "ADD_NOTIFICATION",
  },
  Object {
    "type": "SAVE_SUCCEEDED",
  },
]
`;

exports[`Actions: App creates correct PARSE_SNIPPET action object 1`] = `
Object {
  "meta": Object {
    "WebWorker": true,
  },
  "payload": Object {
    "snippet": "Grasssss tastes bad",
  },
  "type": "PARSE_SNIPPET",
}
`;

exports[`Actions: App creates correct RESET_RULE_FILTERS action object 1`] = `
Object {
  "type": "RESET_RULE_FILTERS",
}
`;

exports[`Actions: App creates correct RESET_STATE action object 1`] = `
Object {
  "type": "RESET_STATE",
}
`;

exports[`Actions: App creates correct RESET_STATE action object 2`] = `
Object {
  "payload": Object {
    "begin": 0,
    "children": Array [
      "NEWLINE",
      "INDENT",
      "NEWLINE",
      "DEDENT",
    ],
    "end": 1,
    "type": "file_input",
  },
  "type": "SET_AST",
}
`;

exports[`Actions: App creates correct RESTORE_STATE action object 1`] = `
Object {
  "payload": Object {
    "AST": Object {
      "begin": 0,
      "children": Array [
        "NEWLINE",
        "INDENT",
        "NEWLINE",
        "DEDENT",
      ],
      "end": 1,
      "type": "file_input",
    },
    "annotations": Object {},
    "filters": Object {},
    "readOnly": false,
    "snippet": "	",
    "snippetTitle": "",
  },
  "type": "RESTORE_STATE",
}
`;

exports[`Actions: App creates correct SAVE_ANNOTATION action object 1`] = `
Object {
  "payload": Object {
    "annotation": "You pass butter.",
    "lineNumber": 1,
    "lineText": "What is my purpose?",
  },
  "type": "SAVE_ANNOTATION",
}
`;

exports[`Actions: App creates correct SELECT_ALL_FILTERS action object 1`] = `
Object {
  "type": "SELECT_ALL_FILTERS",
}
`;

exports[`Actions: App creates correct SET_RULE_FILTERS action object 1`] = `
Object {
  "payload": Object {
    "atom": Object {
      "count": 21,
      "prettyTokenName": "Atoms",
      "selected": true,
    },
    "for_stmt": Object {
      "count": 1,
      "prettyTokenName": "For Loops",
      "selected": false,
    },
  },
  "type": "SET_RULE_FILTERS",
}
`;

exports[`Actions: App creates correct SET_SNIPPET_CONTENTS action object 1`] = `
Object {
  "payload": "Show me what you got",
  "type": "SET_SNIPPET_CONTENTS",
}
`;

exports[`Actions: App creates correct SET_SNIPPET_KEY action object 1`] = `
Object {
  "payload": "szechuan sauce",
  "type": "SET_SNIPPET_KEY",
}
`;

exports[`Actions: App creates correct SET_SNIPPET_LANGUAGE action object 1`] = `
Object {
  "payload": "Gromflomite",
  "type": "SET_SNIPPET_LANGUAGE",
}
`;

exports[`Actions: App creates correct TOGGLE_EDITING_STATE action object 1`] = `
Object {
  "type": "TOGGLE_EDITING_STATE",
}
`;
